src/osp/Active/SysSceneGraph.cpp

```diff
@@ -0,0 +1,211 @@
+                              ? 0
+                              : rScnGraph.m_entToTreePos[std::size_t(parent)];
+
+    uint32_t const  descendants = rScnGraph.m_treeDescendants[parentPos];
+    TreePos_t const childFirst  = parentPos + 1;
+    TreePos_t const childLast   = parentPos + 1 + descendants;
+
+    return ChildRange_t{ChildIterator{&rScnGraph, childFirst},
+                        ChildIterator{&rScnGraph, childLast}};
+}
+
+void SysSceneGraph::do_delete(ACtxSceneGraph& rScnGraph)
+{
+    // Delete subtrees by carefully shifting elements left
+    // Operation being a single left->right swoop. This should be pretty fast
+    // since this is only a few KB in size, and is done once per update
+
+    std::sort(rScnGraph.m_delete.begin(), rScnGraph.m_delete.end());
+
+    TreePos_t const treeLast    = 1 + rScnGraph.m_treeDescendants[0];
+
+    auto const& itTreeDescFirst = rScnGraph.m_treeDescendants.begin();
+    auto const& itTreeEntsFirst = rScnGraph.m_treeToEnt.begin();
+
+    auto const& itDelFirst      = rScnGraph.m_delete.begin();
+    auto const& itDelLast       = rScnGraph.m_delete.end();
+    auto        itDel           = itDelFirst;
+
+    TreePos_t const untouched = (*itDel);
+    TreePos_t done = untouched;
+
+    while (itDel != itDelLast)
+    {
+        auto const itDelNext = std::next(itDel);
+        bool const notLast = (itDelNext != itDelLast);
+
+        uint32_t const removeTotal = 1 + rScnGraph.m_treeDescendants[*itDel];
+
+        // State of array each iteration:
+        //
+        // [Done] [Prev. shifted] [Delete] [Keep] [Delete Next] ....
+        //        <--------SHIFT-----------|----|
+
+        TreePos_t const keepFirst   = (*itDel) + removeTotal;
+        TreePos_t const keepLast    = (notLast) ? (*itDelNext) : (treeLast);
+        assert(keepFirst <= keepLast);
+
+        std::size_t const shift     = keepFirst - done;
+
+        // Update descendant count of ancestors
+        ActiveEnt parent = rScnGraph.m_treeToEnt[*itDel];
+        bool parentNotNull = true;
+        while (parentNotNull)
+        {
+            parent = rScnGraph.m_entParent[std::size_t(parent)];
+            parentNotNull = (parent != lgrn::id_null<ActiveEnt>());
+            TreePos_t const parentPos = parentNotNull ? rScnGraph.m_entToTreePos[std::size_t(parent)] : 0;
+            rScnGraph.m_treeDescendants[parentPos] -= removeTotal;
+        }
+
+        // Clear values for entities to delete
+        std::for_each(itTreeEntsFirst + (*itDel), itTreeEntsFirst + (*itDel + removeTotal),
+                      [&rScnGraph] (ActiveEnt const ent)
+        {
+            rScnGraph.m_entParent[std::size_t(ent)]     = lgrn::id_null<ActiveEnt>();
+            rScnGraph.m_entToTreePos[std::size_t(ent)]  = lgrn::id_null<TreePos_t>();
+        });
+
+        // Update tree positions for elements to shift
+        std::for_each(itTreeEntsFirst + keepFirst, itTreeEntsFirst + keepLast,
```

Comment:
use std::next(itTreeEntsFirst, keepFirst)

Reply:
I started having second thoughts about these. `itTreeEntsFirst + keepFirst` is far more readable than `std::next(itTreeEntsFirst, keepFirst)`. Can't std::next/advance/begin/end be reserved for template functions only, where the iterator types are actually at risk of not having the right operations?